File: examples/python/09_async_usage.py
"""
Async Usage Example

Demonstrates asynchronous operations with ModelSuite.

Requirements:
- Set GROQ_API_KEY environment variable (Groq has fast inference)

Run:
    python 09_async_usage.py
"""

import asyncio
from modelsuite import AsyncModelSuiteClient, Message, CompletionRequest


async def basic_async_completion():
    """Basic async completion request."""
    client = AsyncModelSuiteClient.from_env()

    # Use "provider/model" format for explicit provider routing
    request = CompletionRequest(
        model="groq/llama-3.3-70b-versatile",
        messages=[Message.user("What is the capital of Japan?")],
        max_tokens=100,
    )

    print("Making async request...")
    response = await client.complete(request)
    print(f"Response: {response.text_content()}")

    return response


async def async_streaming():
    """Async streaming responses."""
    client = AsyncModelSuiteClient.from_env()

    request = CompletionRequest(
        model="groq/llama-3.3-70b-versatile",
        messages=[Message.user("Count from 1 to 5 slowly")],
        max_tokens=200,
    ).with_streaming()

    print("Streaming async response:")
    async for chunk in await client.complete_stream(request):
        if chunk.text:
            print(chunk.text, end="", flush=True)
        if chunk.is_done:
            break
    print()


async def concurrent_requests():
    """Make multiple requests concurrently."""
    client = AsyncModelSuiteClient.from_env()

    # Define multiple requests
    questions = [
        "What is Python?",
        "What is Rust?",
        "What is JavaScript?",
    ]

    requests = [
        CompletionRequest(
            model="groq/llama-3.3-70b-versatile",
            messages=[Message.user(q)],
            max_tokens=100,
        )
        for q in questions
    ]

    print("Making concurrent requests...")

    # Run all requests concurrently
    responses = await asyncio.gather(
        *[client.complete(req) for req in requests],
        return_exceptions=True,
    )

    for question, response in zip(questions, responses):
        if isinstance(response, Exception):
            print(f"\nQ: {question}")
            print(f"A: Error - {response}")
        else:
            print(f"\nQ: {question}")
            print(f"A: {response.text_content()[:100]}...")


async def rate_limited_batch():
    """Process requests with rate limiting."""
    client = AsyncModelSuiteClient.from_env()

    questions = [
        "What is 1+1?",
        "What is 2+2?",
        "What is 3+3?",
        "What is 4+4?",
        "What is 5+5?",
    ]

    # Process with semaphore to limit concurrency
    semaphore = asyncio.Semaphore(2)  # Max 2 concurrent requests

    async def limited_request(question):
        async with semaphore:
            print(f"Processing: {question}")
            request = CompletionRequest(
                model="groq/llama-3.3-70b-versatile",
                messages=[Message.user(question)],
                max_tokens=20,
            )
            response = await client.complete(request)
            return question, response.text_content().strip()

    print("Processing with rate limiting (max 2 concurrent)...")
    results = await asyncio.gather(*[limited_request(q) for q in questions])

    print("\nResults:")
    for question, answer in results:
        print(f"  {question} -> {answer}")


async def async_with_timeout():
    """Handle timeouts in async requests."""
    client = AsyncModelSuiteClient.from_env()

    request = CompletionRequest(
        model="groq/llama-3.3-70b-versatile",
        messages=[Message.user("Write a very long story")],
        max_tokens=2000,
    )

    try:
        # Set a timeout
        response = await asyncio.wait_for(
            client.complete(request),
            timeout=30.0,  # 30 second timeout
        )
        print(f"Got response: {response.text_content()[:100]}...")
    except asyncio.TimeoutError:
        print("Request timed out!")


async def async_error_handling():
    """Error handling in async context."""
    from modelsuite import (
        ModelSuiteError,
        AuthenticationError,
        RateLimitError,
    )

    client = AsyncModelSuiteClient.from_env()

    request = CompletionRequest(
        model="groq/llama-3.3-70b-versatile",
        messages=[Message.user("Hello!")],
        max_tokens=100,
    )

    try:
        response = await client.complete(request)
        print(f"Success: {response.text_content()}")
    except AuthenticationError:
        print("Auth error - check API key")
    except RateLimitError as e:
        print(f"Rate limited: {e}")
    except ModelSuiteError as e:
        print(f"ModelSuite error: {e}")
    except Exception as e:
        print(f"Unexpected error: {e}")


async def main():
    print("=" * 50)
    print("Example 1: Basic Async Completion")
    print("=" * 50)
    await basic_async_completion()

    print("\n" + "=" * 50)
    print("Example 2: Async Streaming")
    print("=" * 50)
    await async_streaming()

    print("\n" + "=" * 50)
    print("Example 3: Concurrent Requests")
    print("=" * 50)
    await concurrent_requests()

    print("\n" + "=" * 50)
    print("Example 4: Rate-Limited Batch")
    print("=" * 50)
    await rate_limited_batch()

    print("\n" + "=" * 50)
    print("Example 5: Async Error Handling")
    print("=" * 50)
    await async_error_handling()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/nodejs/09-async-usage.ts
/**
 * Async Usage Example
 *
 * Demonstrates asynchronous patterns with LLMKit in Node.js/TypeScript.
 * Since Node.js is async by nature, this focuses on concurrent execution,
 * rate limiting, and Promise patterns.
 *
 * Requirements:
 * - Set GROQ_API_KEY environment variable (Groq has fast inference)
 * - npm install
 *
 * Run:
 *   npx ts-node 09-async-usage.ts
 */

import {
    JsLLMKitClient as LLMKitClient,
    JsMessage as Message,
    JsCompletionRequest as CompletionRequest,
} from 'llmkit'

async function basicAsyncCompletion() {
    const client = LLMKitClient.fromEnv()

    const request = CompletionRequest
        .create('groq/llama-3.3-70b-versatile', [
            Message.user('What is the capital of Japan?')
        ])
        .withMaxTokens(100)

    console.log('Making async request...')
    const response = await client.complete(request)
    console.log(`Response: ${response.textContent()}`)
}

async function asyncStreaming() {
    const client = LLMKitClient.fromEnv()

    const request = CompletionRequest
        .create('groq/llama-3.3-70b-versatile', [
            Message.user('Count from 1 to 5 slowly')
        ])
        .withMaxTokens(200)
        .withStreaming()

    console.log('Streaming async response:')
    const stream = await client.completeStream(request)
    for await (const chunk of stream) {
        if (chunk.text) {
            process.stdout.write(chunk.text)
        }
        if (chunk.isDone) break
    }
    console.log()
}

async function concurrentRequests() {
    const client = LLMKitClient.fromEnv()

    const questions = [
        'What is Python?',
        'What is Rust?',
        'What is JavaScript?',
    ]

    const requests = questions.map(q =>
        CompletionRequest
            .create('groq/llama-3.3-70b-versatile', [Message.user(q)])
            .withMaxTokens(100)
    )

    console.log('Making concurrent requests...')

    // Run all requests concurrently with Promise.all
    const responses = await Promise.all(
        requests.map(req => client.complete(req).catch(e => e))
    )

    for (let i = 0; i < questions.length; i++) {
        const response = responses[i]
        console.log(`\nQ: ${questions[i]}`)
        if (response instanceof Error) {
            console.log(`A: Error - ${response.message}`)
        } else {
            console.log(`A: ${response.textContent().slice(0, 100)}...`)
        }
    }
}

async function rateLimitedBatch() {
    const client = LLMKitClient.fromEnv()

    const questions = [
        'What is 1+1?',
        'What is 2+2?',
        'What is 3+3?',
        'What is 4+4?',
        'What is 5+5?',
    ]

    // Simple semaphore implementation for rate limiting
    const maxConcurrent = 2
    let running = 0
    const queue: (() => void)[] = []

    async function withLimit<T>(fn: () => Promise<T>): Promise<T> {
        while (running >= maxConcurrent) {
            await new Promise<void>(resolve => queue.push(resolve))
        }
        running++
        try {
            return await fn()
        } finally {
            running--
            const next = queue.shift()
            if (next) next()
        }
    }

    console.log('Processing with rate limiting (max 2 concurrent)...')

    const results = await Promise.all(
        questions.map(q =>
            withLimit(async () => {
                console.log(`Processing: ${q}`)
                const request = CompletionRequest
                    .create('groq/llama-3.3-70b-versatile', [Message.user(q)])
                    .withMaxTokens(20)
                const response = await client.complete(request)
                return { question: q, answer: response.textContent().trim() }
            })
        )
    )

    console.log('\nResults:')
    for (const { question, answer } of results) {
        console.log(`  ${question} -> ${answer}`)
    }
}

async function asyncWithTimeout() {
    const client = LLMKitClient.fromEnv()

    const request = CompletionRequest
        .create('groq/llama-3.3-70b-versatile', [
            Message.user('Write a very long story')
        ])
        .withMaxTokens(2000)

    // Helper function to add timeout to a promise
    function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
        return Promise.race([
            promise,
            new Promise<never>((_, reject) =>
                setTimeout(() => reject(new Error('Request timed out')), ms)
            )
        ])
    }

    try {
        const response = await withTimeout(client.complete(request), 30000)
        console.log(`Got response: ${response.textContent().slice(0, 100)}...`)
    } catch (error) {
        if (error instanceof Error && error.message === 'Request timed out') {
            console.log('Request timed out!')
        } else {
            throw error
        }
    }
}

async function asyncErrorHandling() {
    const client = LLMKitClient.fromEnv()

    const request = CompletionRequest
        .create('groq/llama-3.3-70b-versatile', [Message.user('Hello!')])
        .withMaxTokens(100)

    try {
        const response = await client.complete(request)
        console.log(`Success: ${response.textContent()}`)
    } catch (error) {
        if (error instanceof Error) {
            const message = error.message.toLowerCase()
            if (message.includes('auth') || message.includes('api key')) {
                console.log('Auth error - check API key')
            } else if (message.includes('rate limit')) {
                console.log(`Rate limited: ${error.message}`)
            } else {
                console.log(`LLMKit error: ${error.message}`)
            }
        } else {
            console.log(`Unexpected error: ${error}`)
        }
    }
}

async function main() {
    console.log('='.repeat(50))
    console.log('Example 1: Basic Async Completion')
    console.log('='.repeat(50))
    await basicAsyncCompletion()

    console.log('\n' + '='.repeat(50))
    console.log('Example 2: Async Streaming')
    console.log('='.repeat(50))
    await asyncStreaming()

    console.log('\n' + '='.repeat(50))
    console.log('Example 3: Concurrent Requests')
    console.log('='.repeat(50))
    await concurrentRequests()

    console.log('\n' + '='.repeat(50))
    console.log('Example 4: Rate-Limited Batch')
    console.log('='.repeat(50))
    await rateLimitedBatch()

    console.log('\n' + '='.repeat(50))
    console.log('Example 5: Async Error Handling')
    console.log('='.repeat(50))
    await asyncErrorHandling()
}

main().catch(console.error)
